src/pkcs11/pkcs11-object.c

```diff
@@ -1038,7 +1104,67 @@ CK_RV C_UnwrapKey(CK_SESSION_HANDLE hSession,	/* the session's handle */
 		  CK_ULONG ulAttributeCount,	/* # of attributes in template */
 		  CK_OBJECT_HANDLE_PTR phKey)
 {				/* gets handle of recovered key */
-	return CKR_FUNCTION_NOT_SUPPORTED;
+	CK_RV rv;
+	CK_BBOOL can_unwrap;
+	CK_KEY_TYPE key_type;
+	CK_ATTRIBUTE unwrap_attribute = { CKA_UNWRAP, &can_unwrap, sizeof(can_unwrap) };
+	CK_ATTRIBUTE key_type_attr = { CKA_KEY_TYPE, &key_type, sizeof(key_type) };
+	struct sc_pkcs11_session *session;
+	struct sc_pkcs11_object *object;
+	struct sc_pkcs11_object *key_object;
+
+	if (pMechanism == NULL_PTR)
+		return CKR_ARGUMENTS_BAD;
+
+	rv = sc_pkcs11_lock();
+	if (rv != CKR_OK)
+		return rv;
+
+	rv = get_object_from_session(hSession, hUnwrappingKey, &session, &object);
+	if (rv != CKR_OK) {
+		if (rv == CKR_OBJECT_HANDLE_INVALID)
+			rv = CKR_KEY_HANDLE_INVALID;
+		goto out;
+	}
+	if (object->ops->unwrap_key == NULL_PTR) {
+		rv = CKR_KEY_TYPE_INCONSISTENT;
```

Comment:
shouldn't this check moved to sc_pkcs11_unwrap_operation()?

Reply:
Please explain why would you prefer that? 

I think there's a reason to do this check already at this point. It's a simple way to notice that this key cannot be used for unwrapping at all and there is no point in going further and creating the target key object. In line 1146 sc_create_object_int(hSession, pTemplate, ulAttributeCount, phKey, 0); leads to a complex operation which creates the object to receive the unwrapped key material. The comment is a bit misleading, if the target object is CKA_TOKEN=TRUE it actually creates the object on card. This happens before sc_pcks11_unwrap_operation(). I think it's safer and more efficient to get out earlier.

To explain the logic a little more: 
if (object->ops->unwrap_key == NULL_PTR) checks whether this card and this key type support unwrapping at all, and in the next check, looking for unwrap_attribute, we check if we have permission to do that with the specific key object we are handling.